[AOTInductor] Add config to allow buffer mutation

muchulee8 opened the pull request:
Summary:
Add an additional config to allow buffer mutation.
For data that's greater than 2GB, we would need to set it as read-only, otherwise overflow would occur.
This is a temporary solution since it won't handle cases that requires mutable data greater than 2GB.

Test Plan: Included in commit.

Differential Revision: D57514729




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @ColinPeppler @amjames @desertfire @chauhang

chenyang78 commented on torch/_inductor/codecache.py:
```diff
@@ -1924,12 +1924,13 @@ def _compile_consts_linux(consts: bytes) -> str:
 
-            # .data section is between .text and .bss. When the size of .data is large,
-            # during the linking, the relocation of .text against .bss may overflow.
-            # Rename it to .ldata so that it won't be in between the .text and .bss section
```
nit - we may still need this comment?

muchulee8 committed:
[AOTInductor] Add config to allow buffer mutation (#126584)

Summary:

Add an additional config to allow buffer mutation.
For data that's greater than 2GB, we would need to set it as read-only, otherwise overflow would occur.
This is a temporary solution since it won't handle cases that requires mutable data greater than 2GB.

Test Plan: Included in commit.

Differential Revision: D57514729

desertfire commented on torch/_inductor/config.py:
```diff
@@ -744,6 +744,9 @@ class aot_inductor:
     force_mmap_weights: bool = False
 
+    # flag to allow buffer mutation. This would remove the read-only property from buffers.
+    allow_buffer_mutation: bool = False
```
Can we do like `"0" if is_fbcode() else "1"`? (search for other examples in this file)

muchulee8 replied:
I don't think this is the right thing, it's not only about fbcode.
If you take a look at the log and previous comment Intel folks made, they moved from .data to .ldata because it overflows to .bss. The overflow is already happening with cases in OSS world. It only happens that the constants they have fits within .ldata so there's no error. If larger constants comes in again, we will get more failures.

desertfire replied:
I agree that diverging the behavior is not a good idea. How about this: instead of adding a config here, Inductor can detect if the model does buffer mutation (https://github.com/pytorch/pytorch/blob/655038687afd19a4a4c9371b77ff046fd6c84be1/torch/_inductor/lowering.py#L5078C25-L5078C41), then decides what section to use and add const_size checking to explicitly fail when size is over limit. We will work on a more comprehensive support in future, but this at least will notice user with a clear error msg.
